From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Iooss <nicolas.iooss_linux@m4x.org>
Date: Sat, 8 Oct 2016 14:56:19 +0200
Subject: [PATCH] {PLUGIN} Add string ncmp function checker

This plugins helped to detect the following bugs:
- 2e57069c869f ("ALSA: bebob: compare whole string for model name so
  that readers are not confused")
- 31b239824ece ("ath10k: use the right length of "background"")
- 2d9a02744f5a ("nvdimm: use the right length of "pmem"")

It has also detected some bugs reported and fixed by other people:
- 99243b922c9d ("dm table: fix "nvme" test")
---
 scripts/Makefile.gcc-plugins                  |   2 +
 scripts/gcc-plugins/Kconfig                   |   6 +
 scripts/gcc-plugins/string_ncmp_size_plugin.c | 239 ++++++++++++++++++
 sound/soc/intel/boards/bxt_rt298.c            |   9 +-
 4 files changed, 250 insertions(+), 6 deletions(-)
 create mode 100644 scripts/gcc-plugins/string_ncmp_size_plugin.c

diff --git a/scripts/Makefile.gcc-plugins b/scripts/Makefile.gcc-plugins
index 40efdfbb46fb..46211a40b50c 100644
--- a/scripts/Makefile.gcc-plugins
+++ b/scripts/Makefile.gcc-plugins
@@ -17,6 +17,8 @@ endif
 
 gcc-plugin-$(CONFIG_GCC_PLUGIN_DEREF_CHECKER) += deref_checker_plugin.so
 
+gcc-plugin-$(CONFIG_GCC_PLUGIN_STRING_NCMP_SIZE) += string_ncmp_size_plugin.so
+
 gcc-plugin-$(CONFIG_GCC_PLUGIN_SANCOV)		+= sancov_plugin.so
 
 gcc-plugin-$(CONFIG_GCC_PLUGIN_STRUCTLEAK)	+= structleak_plugin.so
diff --git a/scripts/gcc-plugins/Kconfig b/scripts/gcc-plugins/Kconfig
index 44a7f599d459..71d32cbb1a92 100644
--- a/scripts/gcc-plugins/Kconfig
+++ b/scripts/gcc-plugins/Kconfig
@@ -117,4 +117,10 @@ config GCC_PLUGIN_DEREF_CHECKER
 	help
 	  Check function calls involving pointers about some consistencies.
 
+config GCC_PLUGIN_STRING_NCMP_SIZE
+	bool "Check calls to strncmp-like functions"
+	depends on GCC_PLUGINS
+	help
+	  Check that calls like strncmp(x, "test", 4) are consistent.
+
 endif
diff --git a/scripts/gcc-plugins/string_ncmp_size_plugin.c b/scripts/gcc-plugins/string_ncmp_size_plugin.c
new file mode 100644
index 000000000000..f455997aff7f
--- /dev/null
+++ b/scripts/gcc-plugins/string_ncmp_size_plugin.c
@@ -0,0 +1,239 @@
+/*
+ * Copyright 2016-2018 by Nicolas Iooss
+ * Licensed under the GPL v2, or (at your option) v3
+ *
+ * Check that calls like strncmp(x, "test", 4) are consistent in that the size
+ * is the length of the second argument.
+ */
+
+#include "gcc-common.h"
+
+int plugin_is_GPL_compatible;
+
+static struct plugin_info string_ncmp_size_plugin_info = {
+	.version	= "20190517",
+	.help		= "String n-copy functions size checker\n",
+};
+
+/**
+ * Identifier match helpers
+ */
+static bool identifier_is(const_tree identifier, const char *str)
+{
+	size_t len = strlen(str);
+
+	if (!identifier)
+		return false;
+
+	return IDENTIFIER_LENGTH(identifier) == len && !strncmp(str, IDENTIFIER_POINTER(identifier), len);
+}
+
+/*
+ * Extract the length from a tree if it is constant
+ */
+static bool extract_constant_length(const_tree tree, HOST_WIDE_INT *plen)
+{
+	gcc_assert(TREE_CODE(TREE_TYPE(tree)) == INTEGER_TYPE);
+	if (!CONSTANT_CLASS_P(tree))
+		return false;
+	*plen = TREE_INT_CST_ELT_CHECK(tree, 0);
+	return true;
+}
+
+/* Function from Emese's initify_plugin
+ * https://github.com/ephox-gcc-plugins/initify/blob/master/initify_plugin.c
+ */
+static const_tree get_string_cst(const_tree var)
+{
+	if (var == NULL_TREE)
+		return NULL_TREE;
+
+	if (TREE_CODE(var) == STRING_CST)
+		return var;
+
+	switch (TREE_CODE_CLASS(TREE_CODE(var))) {
+	case tcc_expression:
+	case tcc_reference: {
+		int i;
+
+		for (i = 0; i < TREE_OPERAND_LENGTH(var); i++) {
+			const_tree ret = get_string_cst(TREE_OPERAND(var, i));
+
+			if (ret != NULL_TREE)
+				return ret;
+		}
+		break;
+	}
+
+	default:
+		break;
+	}
+	return NULL_TREE;
+}
+
+static void analyze_call_statement(gimple stmt)
+{
+	gcall *call_stmt;
+	const_tree fndecl, fnident, arg_src, arg_len;
+	bool is_memcmp, is_strncmp; /* FIXME add memcpy? */
+	HOST_WIDE_INT len_from_arg;
+	int srclen;
+	unsigned int arg_index;
+
+	call_stmt = as_a_gcall(stmt);
+
+	/* Get the function declaration */
+	fndecl = gimple_call_fndecl(call_stmt);
+	if (fndecl == NULL_TREE)
+		return;
+
+	/* Filter functions by name */
+	/* FIXME instead of booleans, let's categorize the parameter types in a global dictionary */
+	/* FIXME same thing for strdup/memdup ? */
+	fnident = DECL_NAME(fndecl);
+	is_memcmp = (identifier_is(fnident, "__builtin_memcmp") ||
+		     identifier_is(fnident, "__underlying_memcmp") ||
+		     identifier_is(fnident, "memcmp") ||
+		     identifier_is(fnident, "reverse_memcmp") ||
+		     identifier_is(fnident, "stackdepot_memcmp"));
+	is_strncmp = (identifier_is(fnident, "strncmp") ||
+		      identifier_is(fnident, "ucs2_strncmp"));
+	if (!is_memcmp && !is_strncmp) {
+		/* Check that it is not a function which would be interesting
+		 * Here is the list of known functions which are not.
+		 */
+		if (!identifier_is(fnident, "kasan_memcmp") &&
+		    !identifier_is(fnident, "memcmp_extent_buffer") &&
+		    !identifier_is(fnident, "memcmp_node_keys") &&
+		    !identifier_is(fnident, "memcmp_pages") &&
+		    !identifier_is(fnident, "memcmpb") &&
+		    !identifier_is(fnident, "memcmpshowoffset") &&
+		    !identifier_is(fnident, "masked_memcmp") && /*  FIXME may be interesting to adapt the plugin for 3 arguments */
+		    !identifier_is(fnident, "nla_memcmp") && /* FIXME may be interesting to adapt the plugin for (vect, size) functions */
+		    !identifier_is(fnident, "nvbios_memcmp") && /* FIXME may be interesting to adapt the plugin for (vect, size) functions */
+		    !identifier_is(fnident, "utf8_strncmp") &&
+		    !identifier_is(fnident, "UniStrncmp_le")) {
+			size_t len = IDENTIFIER_LENGTH(fnident);
+			const char *name = IDENTIFIER_POINTER(fnident);
+			gcc_assert(name[len] == '\0');
+			if (strcasestr(name, "memcmp") || strcasestr(name, "strncmp")) {
+				warning_at(
+					gimple_location(stmt), 0,
+					G_("Call to %qF may be interesting to analyze in the string length analyzer plugin. If not, please add it to %s.\n"),
+					fndecl, __FILE__);
+			}
+		}
+		return;
+	}
+
+	/* Retreive the length as a constant integer */
+	arg_len = gimple_call_arg(call_stmt, 2);
+	if (TREE_CODE(TREE_TYPE(arg_len)) != INTEGER_TYPE && TREE_CODE(TREE_TYPE(arg_len)) != ENUMERAL_TYPE) {
+		warning_at(
+			gimple_location(stmt), 0,
+			G_("Invalid call to %qF: non-integer size\n"),
+			fndecl);
+		return;
+	}
+	if (!CONSTANT_CLASS_P(arg_len))
+		return;
+	len_from_arg = TREE_INT_CST_ELT_CHECK(arg_len, 0);
+
+	/* Retrieve the const string in the second argument */
+	arg_index = 1;
+	arg_src = gimple_call_arg(call_stmt, arg_index);
+	arg_src = get_string_cst(arg_src);
+	if (arg_src == NULL_TREE) {
+		/* Sometimes the constant if the first argument */
+		arg_index = 0;
+		arg_src = gimple_call_arg(call_stmt, arg_index);
+		arg_src = get_string_cst(arg_src);
+		if (arg_src == NULL_TREE) {
+			/* Do not do anything for calls with two variables */
+			return;
+		}
+	}
+	gcc_assert(TREE_CODE(arg_src) == STRING_CST);
+	srclen = TREE_STRING_LENGTH(arg_src) - 1;
+	if (srclen <= 0) {
+		warning_at(
+			gimple_location(stmt), 0,
+			G_("Invalid call to %qF: empty constant string\n"),
+			fndecl);
+		return;
+	}
+	/* memcmp can use \0 in its strings but not strncmp */
+	if (is_strncmp && strnlen(TREE_STRING_POINTER(arg_src), srclen) != srclen) {
+		warning_at(
+			gimple_location(stmt), 0,
+			G_("Invalid call to %qF: string with nul characters\n"),
+			fndecl);
+	}
+
+	/* Compare things: prefix or full string */
+	if (len_from_arg == srclen || len_from_arg == srclen + 1) {
+		/* All right */
+		return;
+	}
+	/* Allow strncmp(vec, "string", sizeof(vec)) where vec is a char[] */
+	if (is_strncmp) {
+		const_tree arg_vec;
+		HOST_WIDE_INT veclen;
+
+		arg_vec = gimple_call_arg(call_stmt, 1 - arg_index);
+		if (TREE_CODE(TREE_TYPE(arg_vec)) == POINTER_TYPE &&
+		    TREE_CODE(TREE_TYPE(TREE_TYPE(arg_vec))) == ARRAY_TYPE) {
+			arg_vec = TREE_TYPE(TREE_TYPE(arg_vec));
+			veclen = TREE_INT_CST_ELT_CHECK(TYPE_SIZE_UNIT(arg_vec), 0);
+			if (veclen == len_from_arg)
+				return;
+		}
+	}
+	warning_at(
+		gimple_location(stmt), 0,
+		G_("Invalid call to %qF: length %"HOST_WIDE_INT_PRINT"d but string size %lu\n"),
+		fndecl, len_from_arg, srclen);
+}
+
+static unsigned int string_ncmp_size_execute(void)
+{
+	basic_block bb;
+	gimple_stmt_iterator gsi;
+	gimple stmt;
+
+	FOR_EACH_BB_FN(bb, cfun) {
+		for (gsi = gsi_start_bb(bb); !gsi_end_p(gsi); gsi_next(&gsi)) {
+			stmt = gsi_stmt(gsi);
+			if (is_gimple_call(stmt))
+				analyze_call_statement(stmt);
+		}
+	}
+
+	return 0;
+}
+
+#define PASS_NAME string_ncmp_size
+
+#define NO_GATE
+#define TODO_FLAGS_FINISH TODO_dump_func
+
+#include "gcc-generate-gimple-pass.h"
+
+int plugin_init(struct plugin_name_args *plugin_info, struct plugin_gcc_version *version)
+{
+	const char * const plugin_name = plugin_info->base_name;
+
+	PASS_INFO(string_ncmp_size, "ssa", 1, PASS_POS_INSERT_AFTER);
+
+	if (!plugin_default_version_check(version, &gcc_version)) {
+		error(G_("incompatible gcc/plugin versions"));
+		return 1;
+	}
+
+	register_callback(plugin_name, PLUGIN_INFO, NULL,
+				&string_ncmp_size_plugin_info);
+	register_callback(plugin_name, PLUGIN_PASS_MANAGER_SETUP, NULL,
+				&string_ncmp_size_pass_info);
+
+	return 0;
+}
diff --git a/sound/soc/intel/boards/bxt_rt298.c b/sound/soc/intel/boards/bxt_rt298.c
index 7a4decf34191..aa3b6626137f 100644
--- a/sound/soc/intel/boards/bxt_rt298.c
+++ b/sound/soc/intel/boards/bxt_rt298.c
@@ -603,14 +603,11 @@ static int broxton_audio_probe(struct platform_device *pdev)
 	int i;
 
 	for (i = 0; i < ARRAY_SIZE(broxton_rt298_dais); i++) {
-		if (!strncmp(card->dai_link[i].codecs->name, "i2c-INT343A:00",
-			     I2C_NAME_SIZE)) {
-			if (!strncmp(card->name, "broxton-rt298",
-				     PLATFORM_NAME_SIZE)) {
+		if (!strcmp(card->dai_link[i].codecs->name, "i2c-INT343A:00")) {
+			if (!strcmp(card->name, "broxton-rt298")) {
 				card->dai_link[i].name = "SSP5-Codec";
 				card->dai_link[i].cpus->dai_name = "SSP5 Pin";
-			} else if (!strncmp(card->name, "geminilake-rt298",
-					    PLATFORM_NAME_SIZE)) {
+			} else if (!strcmp(card->name, "geminilake-rt298")) {
 				card->dai_link[i].name = "SSP2-Codec";
 				card->dai_link[i].cpus->dai_name = "SSP2 Pin";
 			}
-- 
